### PR TITLE
Refactor Delete Files to handle Single Responsibility and Added Unit test Cases.

### LINKED ISSUE #4917: Add edge cases for `DeleteFiles` side effect.
* We already have the test case for `DeleteFiles`. But it does not cover all the logic of this class. We should add edge cases for this class.

See https://app.codecov.io/gh/kiwix/kiwix-android/blob/main/app%2Fsrc%2Fmain%2Fjava%2Forg%2Fkiwix%2Fkiwixmobile%2FzimManager%2FfileselectView%2Feffects%2FDeleteFiles.kt for more details.

* Also, use the DI-provided `IODispatcher` in this class.

### PR DESCRIPTION
**Pull request by atharvyadav22**
Fixes: #4917

- Created DeleteFilesUseCase to manage Buisness logic
- Added Test Cases for DeleteFileUseCase

For Copilot review (https://github.com/kiwix/kiwix-android/pull/4941#discussion_r3392066806) This Extension function `Files.isExists` will be covered in https://github.com/kiwix/kiwix-android/issues/4936 as said many of the files depends on this extension function so keeping this PR isolated from Dispatcher.

### CODE REVIEW

**Comment by atharvyadav22** [comment]
@gouri-panda @MohitMaliFtechiz 
I have chaged the design have a look its more testable.

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    file?.let {
+      FileUtils.deleteZimFile(it.path)
+
+      if (it.isFileExist()) {
+        return@deleteBook false
+      }
+    }
+
+    libkiwixBookOnDisk.delete(book.book.id)
+    return true
```
Comment (gouri-panda) [inline app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt:58]: If `book.zimReaderSource.file` is null the `let ` block skipped entirely and the function falls through to deleting the DB record and returning true. The original code had the same bug.

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt [hunk 2]
```diff
@@ -19,77 +19,59 @@
...
-  SideEffect<Unit> {
-  @Inject lateinit var libkiwixBookOnDisk: LibkiwixBookOnDisk
-
-  @Inject lateinit var zimReaderContainer: ZimReaderContainer
-
-  @Suppress("InjectDispatcher")
+  private val dialogShower: DialogShower,
+  private val deleteFilesUseCase: DeleteFilesUseCase,
+  private val viewModelScope: CoroutineScope,
+  private val ioDispatcher: CoroutineDispatcher
```
Comment (gouri-panda) [inline app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt:38]: This constructor takes `viewModelScope`. It's a `SideEffect` short lived data object. Passing `viewModelScope` into it and launching from there fine at time but the old code used `activity.lifecycleScope` which was tied to the UI lifecycle. If we're switching from `viewModelScope` means if the activity is destroyed mid the coroutine will keep running.

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt [hunk 3]
```diff
@@ -19,77 +19,59 @@
...
 import org.kiwix.kiwixmobile.core.extensions.toast
-import org.kiwix.kiwixmobile.core.reader.ZimReaderContainer
 import org.kiwix.kiwixmobile.core.utils.dialog.DialogShower
 import org.kiwix.kiwixmobile.core.utils.dialog.KiwixDialog.DeleteZims
-import org.kiwix.kiwixmobile.core.utils.files.FileUtils
 import org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.BooksOnDiskListItem.BookOnDisk
-import javax.inject.Inject
 
-data class DeleteFiles(
+class DeleteFiles(
```
Comment (gouri-panda) [inline app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt]: I don't think this is ok. The `SideEffect` comparisons in ViewModel tests used for structural equality. @MohitMaliFtechiz Is this ok?

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt (hunk 2 above)
Comment (atharvyadav22) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt:38]: yes but since delete is an transactional operation so using lifecycleScope can stop midway this process.

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt (hunk 1 above)
Comment (atharvyadav22) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt:58]: yes great observation i agree with this

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt (hunk 3 above)
Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt]: @gouri-panda Agree with you. It is not okay to change it. 


@atharvyadav22 Please see and make the changes suggested by @gouri-panda and @Copilot.

Comment (atharvyadav22) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt]: @gouri-panda 
but data class is not useful as its SideEffects and we are not using data class behaviour here
If you suggest i'll reapply again.

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt (hunk 1 above)
Comment (atharvyadav22) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt:58]: @gouri-panda 
I will fix this is there any other portion i should fix?

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt (hunk 3 above)
Comment (gouri-panda) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt]: @atharvyadav22 I'm relying on the @MohitMaliFtechiz decisions here. @MohitMaliFtechiz what's the best thing to do?
>but data class is not useful as its SideEffects and we are not using data class behaviour here

Then why it was in the first place?

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt (hunk 2 above)
Comment (gouri-panda) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt:38]: >yes but since delete is an transactional operation so using lifecycleScope can stop midway this process.

Very good point! thanks!

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt (hunk 1 above)
Comment (gouri-panda) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt:58]: Is there? I don't know. Probably @MohitMaliFtechiz may find it.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/local/LocalLibraryViewModel.kt [hunk 4]
```diff
@@ -321,7 +323,9 @@ class LocalLibraryViewModel @Inject constructor(
     when (action) {
       is RequestMultiSelection -> noSideEffectSelectBook(action.bookOnDisk)
       is RequestSelect -> noSideEffectSelectBook(action.bookOnDisk)
-      RequestDeleteMultiSelection -> DeleteFiles(selectionsFromState(), alertDialogShower)
+      RequestDeleteMultiSelection ->
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/local/LocalLibraryViewModel.kt:328]: Fix white space issues.

Image

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+  private suspend fun deleteBook(
+    book: BooksOnDiskListItem.BookOnDisk
+  ): Boolean {
+    val file = book.zimReaderSource.file
+
+    file?.let {
+      FileUtils.deleteZimFile(it.path)
+
+      if (it.isFileExist()) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt]: We should here apply the @Copilot suggestion. https://github.com/kiwix/kiwix-android/pull/4941/changes#r3392066806

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt (hunk 1 above)
Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt:58]: @gouri-panda , @atharvyadav22 This is the only issue i am seeing.

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package org.kiwix.kiwixmobile.zimManager.fileselectView.effects
+
+import jakarta.inject.Inject
```
[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt]: Fix this issue.

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt (hunk 3 above)
Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt]: > Then why it was in the first place?

My assumption is that it was originally made a data class so emitted side effects could be compared directly in tests using structural equality. While the class itself mainly encapsulates behavior, the generated equality semantics are still useful for testing, so keeping it as a data class makes sense for consistency with the existing test setup.

> @atharvyadav22 I'm relying on the @MohitMaliFtechiz decisions here. @MohitMaliFtechiz what's the best thing to do?

All `SideEffect` classes are currently implemented as data classes. For consistency, we should keep this as a data class for now and only consider changing it if we decide to update all SideEffect classes across the codebase.

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt (hunk 5 above)
Comment (atharvyadav22) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt]: So this will be covered in  #4936 as i said many of the files depends on this extension function so keeping this PR isolated from Dispatcher and would file a PR soon.

**Comment by atharvyadav22** [comment]
Done! as per the changes.

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt (hunk 5 above)
Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCase.kt]: @atharvyadav22 You can mention these things in PR description so that it will be easy to track anyone.

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt (hunk 3 above)
Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt]: This fix is missing. Please revert this change for now.

Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFiles.kt]: There is a method in `KiwixActivityComponent`.

```
fun inject(deleteFiles: DeleteFiles)
```

It is not used now. We should remove that.

File: app/src/test/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesTest.kt [hunk 7]
```diff
@@ -20,116 +20,144 @@ package org.kiwix.kiwixmobile.zimManager.fileselectView.effects
...
+
+    deleteFiles =
+      DeleteFiles(
+        booksOnDiskListItems,
+        dialogShower,
+        deleteFilesUseCase,
+        viewModelScope,
+        ioDispatcher.dispatcher
+      )
+    every { activity.toast(any<Int>(), any()) } just Runs
```
[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply app/src/test/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesTest.kt]: Implement this suggestion.

File: app/src/test/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesTest.kt [hunk 8]
```diff
@@ -20,116 +20,144 @@ package org.kiwix.kiwixmobile.zimManager.fileselectView.effects
...
+      dialogShower.show(
+        KiwixDialog.DeleteZims("Book 1\nBook 2"),
+        any()
+      )
+    }
+  }
 
-    val dialogSlot = slot<KiwixDialog>()
+  @Test
+  fun invokeWith_whenDeleteClicked_callsDeleteFilesUseCase() {
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesTest.kt]: @atharvyadav22 When I am running all the tests together, this test is passing. But when I am running this single test, it is failing. So something is missing in this test.

Image

```kotlin

Verification failed: call 1 of 1: DeleteFilesUseCase(#2).invoke(eq([BookOnDisk(databaseId=0, book=LibkiwixBook(nativeBook=null, _id=book-id-1, _title=Book 1, _description=null, _language=, _creator=, _publisher=, _date=, _url=null, _articleCount=null, _mediaCount=null, _size=, _bookName=null, _favicon=, _tags=null, _path=null, searchMatches=0, file=/storage/kiwix.zim), file=, zimReaderSource=org.kiwix.kiwixmobile.core.reader.ZimReaderSource@8c5e315, tags=[], isSelected=false, id=book-id-1), BookOnDisk(databaseId=0, book=LibkiwixBook(nativeBook=null, _id=book-id-2, _title=Book 2, _description=null, _language=, _creator=, _publisher=, _date=, _url=null, _articleCount=null, _mediaCount=null, _size=, _bookName=null, _favicon=, _tags=null, _path=null, searchMatches=0, file=/storage/test.zim), file=, zimReaderSource=org.kiwix.kiwixmobile.core.reader.ZimReaderSource@2817e705, tags=[], isSelected=false, id=book-id-2)]), any())) was not called
java.lang.AssertionError: Verification failed: call 1 of 1: DeleteFilesUseCase(#2).invoke(eq([BookOnDisk(databaseId=0, book=LibkiwixBook(nativeBook=null, _id=book-id-1, _title=Book 1, _description=null, _language=, _creator=, _publisher=, _date=, _url=null, _articleCount=null, _mediaCount=null, _size=, _bookName=null, _favicon=, _tags=null, _path=null, searchMatches=0, file=/storage/kiwix.zim), file=, zimReaderSource=org.kiwix.kiwixmobile.core.reader.ZimReaderSource@8c5e315, tags=[], isSelected=false, id=book-id-1), BookOnDisk(databaseId=0, book=LibkiwixBook(nativeBook=null, _id=book-id-2, _title=Book 2, _description=null, _language=, _creator=, _publisher=, _date=, _url=null, _articleCount=null, _mediaCount=null, _size=, _bookName=null, _favicon=, _tags=null, _path=null, searchMatches=0, file=/storage/test.zim), file=, zimReaderSource=org.kiwix.kiwixmobile.core.reader.ZimReaderSource@2817e705, tags=[], isSelected=false, id=book-id-2)]), any())) was not called
	at io.mockk.impl.recording.states.VerifyingState.failIfNotPassed(VerifyingState.kt:61)
	at io.mockk.impl.recording.states.VerifyingState.recordingDone(VerifyingState.kt:40)
	at io.mockk.impl.recording.CommonCallRecorder.done(CommonCallRecorder.kt:47)
	at io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:63)
	at io.mockk.impl.eval.VerifyBlockEvaluator.verify(VerifyBlockEvaluator.kt:30)
	at io.mockk.MockKDsl.internalCoVerify(API.kt:145)
	at io.mockk.MockKKt.coVerify(MockK.kt:244)
	at io.mockk.MockKKt.coVerify$default(MockK.kt:235)

```

File: app/src/test/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCaseTest.kt [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    coEvery {
+      file1.isFileExist()
+    } returns false
+    coEvery {
+      file2.isFileExist(any())
+    } returns false
+
+    coEvery {
+      FileUtils.deleteZimFile(file1.path)
+    } just Runs
```
[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply app/src/test/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCaseTest.kt]: Apply this suggestion as well.

File: app/src/test/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCaseTest.kt [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    book = BookOnDisk(book = libkiwixBook, zimReaderSource = ZimReaderSource(file1))
+
+    deleteFilesUseCase =
+      DeleteFilesUseCase(libkiwixBookOnDisk, zimReaderContainer)
+  }
+
+  @AfterEach
+  fun tearDown() {
+    clearAllMocks()
+  }
```
[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply app/src/test/java/org/kiwix/kiwixmobile/zimManager/fileselectView/effects/DeleteFilesUseCaseTest.kt:81]: This one too.

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz @gouri-panda 
All issues resolved.

**Review by MohitMaliFtechiz — COMMENTED** [review]
@atharvyadav22 LGTM. Can you please rebase and remove the extra commit?

**Review by gouri-panda — DISMISSED** [review]
@atharvyadav22 Thanks! I still don't see no tests for if the path is null :)

**Comment by atharvyadav22** [comment]
> @atharvyadav22 Thanks! I still don't see no tests for if the path is null :)

I think you are talking about this PR have a look https://github.com/kiwix/kiwix-android/pull/4942#discussion_r3401603213

**Review by MohitMaliFtechiz — DISMISSED** [review]
@atharvyadav22 Thank you.

**Comment by gouri-panda** [comment]
@atharvyadav22 Ooh sorry! I thought i commented in this PR. Its a nice test to add this here too.

**Comment by atharvyadav22** [comment]
> @atharvyadav22 Ooh sorry! I thought i commented in this PR. Its a nice test to add this here too.

@gouri-panda @MohitMaliFtechiz 
Have added an case!

**Review by MohitMaliFtechiz — APPROVED** [review]

**Review by gouri-panda — APPROVED** [review]
@atharvyadav22 Thanks :)
